### PR TITLE
ALL: improve support for temporary command line parameters

### PR DESCRIPTION
**Pull request by grisenti**
Improved support for command line supplied temporary values. These include: 
- [x] -c, --config=CONFIG
- [x] -f, --fullscreen
- [x] -F, --no-fullscreen
- [x] -g, --gfx-mode=MODE 
- [x] --stretch-mode=MODE
- [x] --scaler=MODE
- [x] --scale-factor=FACTOR   
- [x] --filtering              
- [x] --no-filtering          
- [x] --gui-theme=THEME      
- [x] --themepath=PATH       
- [x] -m, --music-volume=NUM   
- [x] -s, --sfx-volume=NUM     
- [x] -r, --speech-volume=NUM  
- [x] --midi-gain=NUM        
- [x] -n, --subtitles         
- [x] --savepath=PATH          
- [x] --extrapath=PATH      
- [x] --soundfont=FILE        
- [x] --multi-midi            
- [x] --native-mt32          
- [x] --enable-gs              
- [x] --opl-driver=DRIVER      
- [x] --talkspeed=NUM          
- [x] --render-mode=MODE

### CODE REVIEW

File: base/main.cpp [hunk 1]
```diff
@@ -427,6 +427,10 @@ extern "C" int scummvm_main(int argc, const char * const argv[]) {
 		ConfMan.loadDefaultConfigFile();
 	}
 
+	if(settings.contains("savepath")){
```
Comment (bluegr) [inline base/main.cpp:430]:
```suggestion
	if (settings.contains("savepath")){
```

File: common/config-manager.cpp [hunk 2]
```diff
@@ -160,6 +162,7 @@ void ConfigManager::loadFromStream(SeekableReadStream &stream) {
 	_miscDomains.clear();
 	_transientDomain.clear();
 	_domainSaveOrder.clear();
+	//_sessionDomain.clear();
```
Comment (bluegr) [inline common/config-manager.cpp:165]: Leftover?

Comment (grisenti) [reply common/config-manager.cpp:165]: Yes, sorry. I didn't know if it was appropriate to clear it or not and in the end I forgot to remove the comment.

File: base/main.cpp [hunk 3]
```diff
@@ -427,6 +427,10 @@ extern "C" int scummvm_main(int argc, const char * const argv[]) {
...
+		ConfMan.set("savepath", settings["savepath"], Common::ConfigManager::kSessionDomain); 
+	}
+
```
Comment (criezy) [inline base/main.cpp:433]: I am wondering if it would not be better to have this toward the end of the `processSettings()` function. Currently this is where all the settings are copied to the transient domain and it seems to me it would make sense to have the copy to the session domain in the same place.

File: common/config-manager.cpp [hunk 4]
```diff
@@ -39,6 +39,7 @@ DECLARE_SINGLETON(ConfigManager);
 
 char const *const ConfigManager::kApplicationDomain = "scummvm";
 char const *const ConfigManager::kTransientDomain = "__TRANSIENT";
+char const *const ConfigManager::kSessionDomain = "session";
```
Comment (criezy) [inline common/config-manager.cpp:42]: Maybe call it `"__SESSION" to avoid a clash with a potential game id that might be called "session"?

File: gui/options.cpp [hunk 5]
```diff
@@ -2116,7 +2116,9 @@ void GlobalOptionsDialog::build() {
 	Common::String iconPath(ConfMan.get("iconspath", _domain));
 	Common::String extraPath(ConfMan.get("extrapath", _domain));
 
-	if (savePath.empty() || !ConfMan.hasKey("savepath", _domain)) {
+	if (ConfMan.hasKey("savepath", Common::ConfigManager::kSessionDomain)){
+		_savePath->setLabel(_("Command line temporary"));
```
Comment (criezy) [inline gui/options.cpp:2120]: I think it could be nice to still display the path, but maybe add a ` (temporary)` suffix to it.

**Review by criezy — CHANGES_REQUESTED** [review]
Please fix the commit message to follow our convention: https://wiki.scummvm.org/index.php?title=Commit_Guidelines#Commit_message_formatting

**Comment by criezy** [comment]
Also a note to my future self: we should use this new session domain for the screenshotpath as well and revert f3dc1df7d. 

And there are probably other settings for which it would make sense as well (such as the log path or config file path, even though it would not change anything for those with the way they are currently used - being accessed only once when starting ScummVM).

File: gui/options.cpp (hunk 5 above)
Comment (grisenti) [reply gui/options.cpp:2120]: I agree it looks better this way

**Comment by criezy** [comment]
That looks OK to me now, except for the commit messages that still do not follow our commit message formatting conventions.

I would also prefer to split the changes into two commits: one for the changes to the ConfigManager (add the new session domain) and another one for the changes related to the savepath command line option.

If you want I can do that when merging.

File: common/config-manager.cpp [hunk 6]
```diff
@@ -405,13 +412,17 @@ ConfigManager::Domain *ConfigManager::getDomain(const String &domName) {
...
-	// 3) the application domain.
+	// 2) the session domain, 
+	// 3) the active game domain (if any),
+	// 4) the application domain.
 	// The defaults domain is explicitly *not* checked.
 
 	if (_transientDomain.contains(key))
 		return true;
 
+	if(_sessionDomain.contains(key))
```
Comment (sev-) [inline common/config-manager.cpp:423]: Formatting

File: common/config-manager.cpp [hunk 7]
```diff
@@ -452,6 +463,8 @@ void ConfigManager::removeKey(const String &key, const String &domName) {
 const String &ConfigManager::get(const String &key) const {
 	if (_transientDomain.contains(key))
 		return _transientDomain[key];
+	else if(_sessionDomain.contains(key))
```
Comment (sev-) [inline common/config-manager.cpp:466]: Formatting

File: gui/options.cpp [hunk 8]
```diff
@@ -2116,7 +2116,9 @@ void GlobalOptionsDialog::build() {
 	Common::String iconPath(ConfMan.get("iconspath", _domain));
 	Common::String extraPath(ConfMan.get("extrapath", _domain));
 
-	if (savePath.empty() || !ConfMan.hasKey("savepath", _domain)) {
+	if (ConfMan.hasKey("savepath", Common::ConfigManager::kSessionDomain)) {
```
Comment (sev-) [inline gui/options.cpp:2119]: I do not like this hardcoding. Instead of making it specific for the savepath, I would rather make it generic, for _all_ keys which are specified by the command line. Examples are `-c`, `-p`, `-g`, `-f` etc.

E.g. you may expand `ConfMan.hasKey()` but then add additional signage that those keys come from the command line. The current approach will _modify_ the path value in the GUI, and effectively prevent path having " (temporary)" string in many languages. 

Rather than modifying the string, you could consider, say, modifying its color and adding an explanation to the tooltip.

File: gui/options.cpp [hunk 9]
```diff
@@ -2613,10 +2615,12 @@ void GlobalOptionsDialog::apply() {
 	bool isRebuildNeeded = false;
 
 	Common::U32String savePath(_savePath->getLabel());
-	if (!savePath.empty() && (savePath != _("Default")))
+	if (!savePath.empty() && (savePath != _("Default")) && (!savePath.contains(_(" (temporary)"))))
```
Comment (sev-) [inline gui/options.cpp:2618]: Instead of doing this, I'd recommend introducing another method to ConfMan, like `ConfMan.isKeyTemporary()`.

File: gui/options.cpp [hunk 10]
```diff
@@ -2613,10 +2615,12 @@ void GlobalOptionsDialog::apply() {
...
 		ConfMan.set("savepath", savePath.encode(), _domain);
-	else
+	else if (savePath.empty() || (savePath == _("Default"))){
```
Comment (sev-) [inline gui/options.cpp:2620]: This looks like you're preventing the path from being edited at all if it was specified in the command line. This should not be the case.

**Review by sev- — CHANGES_REQUESTED** [review]
This needs a rework.

The expected behaviour:

- Any flag/value which is specified by the command line, should be covered by this PR in a uniform fashion
- When a value is specified via command line, it should be marked in the GUI as such
- Nevertheless, the value should still be editable
- When a user edits such overridden value, then the fact that it was overridden is cleared, e.g. removing it from the domain like in your case, and the specified value stored in the config file.
- If a user does not edit such overridden value, then it is not stored in the config file.

File: gui/options.cpp (hunk 10 above)
Comment (grisenti) [reply gui/options.cpp:2620]: the path can be edited. That block just resets the savepath if it's the default string or it's empty.

File: gui/options.cpp (hunk 8 above)
Comment (grisenti) [reply gui/options.cpp:2119]: I didn't think about the fact that now a valid path that contains " (temporary)" could not be saved. I will try to implement it with color.  
The other part I don't think I fully understand. Are you asking for a new interface in the config manager to check if a specific command line key has been specified (like maybe an overload of `hasKey` that takes and enum)?

File: base/commandLine.cpp [hunk 11]
```diff
@@ -1669,6 +1669,73 @@ bool processSettings(Common::String &command, Common::StringMap &settings, Commo
 		}
 	}
 
+	// store all session related settings
+	if (settings.contains("savepath")) {
```
Comment (sev-) [inline base/commandLine.cpp:1673]: I'd recommend either refactoring it into a method or use macro, e.g. something like:

```C++
storeSessionSetting("savepath");
storeSessionSetting("themepath");
```
etc

File: base/main.cpp [hunk 12]
```diff
@@ -366,12 +366,6 @@ static void setupGraphics(OSystem &system) {
 
 	system.applyBackendSettings();
 
-	// When starting up launcher for the first time, the user might have specified
-	// a --gui-theme option, to allow that option to be working, we need to initialize
-	// GUI here.
-	// FIXME: Find a nicer way to allow --gui-theme to be working
-	GUI::GuiManager::instance();
```
Comment (sev-) [inline base/main.cpp:373]: Why is this removed? Does `--gui-theme` option still work?

File: gui/ThemeEngine.cpp [hunk 13]
```diff
@@ -1057,7 +1057,8 @@ void ThemeEngine::drawLineSeparator(const Common::Rect &r) {
 	drawDD(kDDSeparator, r);
 }
 
-void ThemeEngine::drawCheckbox(const Common::Rect &r, int spacing, const Common::U32String &str, bool checked, WidgetStateInfo state, bool rtl) {
+void ThemeEngine::drawCheckbox(const Common::Rect &r, int spacing, const Common::U32String &str, bool checked, 
+							   WidgetStateInfo state, bool override, bool rtl) {
```
Comment (sev-) [inline gui/ThemeEngine.cpp:1061]: `override` is a keyword, not a good name for a variable.

File: gui/ThemeEngine.cpp [hunk 14]
```diff
@@ -1400,6 +1402,29 @@ void ThemeEngine::drawText(const Common::Rect &r, const Common::U32String &str,
 		}
 		break;
 
+	case kFontColorOverride: 
+		if (inverted) {
```
Comment (sev-) [inline gui/ThemeEngine.cpp:1406]: Please simplify the code by removing one level of indentaion:

```C++
if (inverted) {
  colorId = ...
  break;
 }
 
 switch (state) {
  ...
  ```

File: gui/options.cpp [hunk 15]
```diff
@@ -2110,35 +2199,23 @@ void GlobalOptionsDialog::build() {
...
-		_themePath->setLabel(themePath);
-	}
-
-	if (iconPath.empty() || !ConfMan.hasKey("iconspath", _domain)) {
-		_iconPath->setLabel(_c("None", "path"));
-	} else {
-		_iconPath->setLabel(iconPath);
-	}
+	const auto setPath = 
+	[&](GUI::StaticTextWidget *const widget, const Common::String& pathType, const Common::U32String &defaultLabel) {
```
Comment (sev-) [inline gui/options.cpp:2203]: please fix formatting `const Common::String &pathType`

File: gui/options.cpp [hunk 16]
```diff
@@ -2612,29 +2690,22 @@ void GlobalOptionsDialog::apply() {
...
-	else
-		ConfMan.removeKey("iconspath", _domain);
-
-	Common::U32String extraPath(_extraPath->getLabel());
-	if (!extraPath.empty() && (extraPath != _c("None", "path")))
-		ConfMan.set("extrapath", extraPath.encode(), _domain);
-	else
-		ConfMan.removeKey("extrapath", _domain);
+	const auto changePath = 
+	[&](GUI::StaticTextWidget *const widget, const Common::String& pathType, const Common::U32String &defaultLabel){
```
Comment (sev-) [inline gui/options.cpp:2694]: Please fix formatting `const Common::String &pathType, const Common::U32String &defaultLabel) {`

wrong space around`&pathType` and missing space before `{`

File: gui/widget.cpp [hunk 17]
```diff
@@ -703,9 +708,13 @@ void CheckboxWidget::setState(bool state) {
 	}
 	sendCommand(_cmd, _state);
 }
+	
+void CheckboxWidget::setOverride(bool override) {
+		_override = override;
```
Comment (sev-) [inline gui/widget.cpp:713]: Wrong indentation, another instance of bad variable name. `override` is a keyword

**Review by sev- — CHANGES_REQUESTED** [review]
Very solid work, thanks!

I did not test it, though

File: base/main.cpp (hunk 12 above)
Comment (grisenti) [reply base/main.cpp:373]: That was needed because GuiManager would be initialized after the transient domain was cleared. Now that's not an issue

**Comment by sev-** [comment]
Could you please rebase it, so we could merge it before release?

**Comment by grisenti** [comment]
@sev- you mean squash everything into one commit?

**Comment by sev-** [comment]
No, I mean rebase on the tip of the master so it is mergeable. Currently, there is a conflict

**Comment by sev-** [comment]
Here is how to do it: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork

**Comment by sev-** [comment]
Git fetch
Git rebase master
Git push -f

**Comment by sev-** [comment]
No, you did merge. You must do rebase

**Comment by sev-** [comment]
Thank you!
